### DISCUSSION TITLE
ROG Ally: Support for command button and back buttons

### DISCUSSION THREAD

**Discussion by pinecones-sx** [seerge/g-helper#772] [Ideas]
I tried using g-helper as a replacement for armoury crate mainly for full button remapping, but it isn't quite doing what I need.

Here are my suggestions:

- Allow full remapping of all hardkeys on the ROG Ally including the command center, armoury crate, back paddles, AND the existing xbox buttons (A, B, start, select, L3, R3, etc).  One of my issues with the button layout is that it makes more sense for me to have the xbox controller start & select buttons where the command center & armoury crate buttons are, so one of my original goals was to swap those two button sets.
- Allow remapping TO XInput keys, eg xbox button, A, B, start, select, L3, R3, D-pad, etc.
- Allow emulation of xbox one & xbox one elite controller output -- this would allow mapping the paddle buttons to xbox elite controller paddle buttons and also to allow mapping of the capture button on xbox one controllers.
- Figure out mapping of the back paddle buttons as separate inputs (I know from a previous discussion that this was an issue).
- I'm not sure if this is a G-Helper issue, but I was having issues using the back paddle buttons at all, especially in combination key presses when mapping via other utilities.  Specifically, I tried using Handheld Companion.

I am going to be uninstalling the app for now, but if you would like me to be available for testing in the future, I'll gladly make myself available.

**Comment by pinecones-sx** [comment]
After further testing, I re-wrote this suggestion.

**Comment by seerge** [comment]
@pinecones-sx hello, g-helper is a tool that replaces AC for laptops ;) Reason main things work on Rog Ally is cause all main BIOS endpoints / USB controls for Aura / etc are literally the same.

Everything that you have listed here is an description of a full-scale app (that will have no use on Asus laptops). I believe there is good alternatives for that already. For me (a single person doing this in a free time), it's not worth time and effort so far :) 

Especially considering that currently Rog Ally is probably something like 0.5% of all g-helper users https://github.com/seerge/g-helper/discussions/707 :)

**Comment by seerge** [comment]
P.S. Try https://www.rewasd.com/

**Reply by TheEnab1er** [reply]
Hey, love your app man.  Made a video about it: https://youtu.be/-UPfsmZxLss.

I was wondering, would it still be a full program just to get the left Command Center button working?  Since the Armoury Crate button is already working (just like on the Asus Laptops), is that something that would be easier to add? 

Asus has sold over 450,000 ROG Ally's so it could be more popular at some point, but I completely understand it not being something you want to spend a lot of time on.   I'd love to know if just that left button though would be easier to add.

Anyway, still an awesome app, I have it on my G15 and my ROG and will always be using going forward.  I hope people coming from my video also donated to you to help you out :)

**Reply by seerge** [reply]
@TheEnab1er hello, thanks for the nice review :)

In latest release (0.105) https://github.com/seerge/g-helper/releases it should be possible to bind Command Center button + Back pedal(s) (they both trigger same "action") in G-Helper to some action or [even custom hotkey](https://github.com/seerge/g-helper#custom-hotkey-actions)

**Reply by TheEnab1er** [reply]
Thank you so much!  And thanks for all the hard work on this app.  You've created something way beyond what Asus has managed to make for the past 10 years.

I looked through the custom hotkeys and I'm assuming there isn't one for the Xbox Guide button?  I was hoping to map that to the left "command center" button to be able to open the steam menu in big picture.  I'm assuming it's way beyond just that, but I haven't messed with programming and was just wondering what your thoughts are on it.  

Thanks again! :)

**Reply by seerge** [reply]
There is generally 2 "types" of keys on Ally, ones that are handled by windows (majority of buttons, ABXY, joystics, etc - including Xbox button) and ones that are "proprietary" ASUS (ROG key, backpedal, Control Center button).

Ones handled by windows can be therefore remapped with any appropriate app.

G-Helper intentionally monitors only ones that can't be handled anyhow w/o it :)

**Reply by TheEnab1er** [reply]
Ahh that makes sense.  Thank you :)

**Reply by seerge** [reply]
Generally speaking you can open app log (from Extra -> log icon) and press each button on Ally. If immediately after pressing a button you see : 

```
8/4/2023 5:08:53 PM: Key: <something>
```
Or 
```
8/3/2023 11:03:49 PM: WMI event <something>
```

It's proprietary ASUS stuff (that can't be handled by windows)

**Reply by TheEnab1er** [reply]
I didn't even see the log button until you mentioned it!   I see where it's registering 165 for both the back buttons, 166 for the command center button and 56 for the amoury crate button.  This is good to know even on Asus laptops going forward, thanks again! :)

Also, I can't seem to find any program that will map the Xbox Guide button to the ROG button.  My thought was, I'll use G-Helper to map something like 0x73 (F4) for the command center button.  Then use a program to map the Xbox Guide button to F4.  However, in something like REWASD, I can map the "Xbox Guide" button to "F4" and it works fine with a physical keyboard, but the 0x73 (F4) doesn't register in REWASD.  So I can hit F4 on the keyboard and it activates the Xbox Guide button, but pressing custom ROG button that is mapped to 0x73(F4) doesn't work.

And when I activate detect controls in REWASD (so you can press a button and see which device it is), nothing shows up when pressing the Command Center, Armoury Crate or Paddles (even in custom with virtual mapping keys assigned).  It's almost like it functions as a ghost keyboard/device and can't be seen by other programs.  So when pressing an assigned button in G-Helper, it's using some keyboard that doesn't show up anywhere (but will still work in windows).  

I've also tried HC which does somewhat work, but it completely messed up my controls to the point uninstalling/re-installing in device manager wouldn't even fix them.  Ended up formatting my OS and don't plan on installing HC ever again.  Steam also doesn't seem to have a way to do it even with Steam Input for Xbox Controllers as well as extended support enabled.

**Reply by seerge** [reply]
I didn't really get what you want, you want to map ROG key (the one opening g-helper by default) to emit some X-Box controller action (XBox Guide button) ?

Guide button seem to be same as just windows WIN button https://support.xbox.com/en-US/help/hardware-network/accessories/mouse-keyboard

**Reply by TheEnab1er** [reply]
Hey, sorry if I explained it poorly.  Yes, if it'd be possible to re-map the "Command Center (left button)" button to the "Xbox" button to be able to bring up the steam menu on the left.  I can't for the life of me figure out why Asus decided to not include the Xbox Button...  If this requires a bunch of work or isn't possible then it's no issue, I'm already happy with how it is and really appreciate it. 

I saw the guide you sent, that's if you plug a keyboard into an Xbox Console, you can hit windows key.  On a windows computer though it brings up the start menu no matter what.   Basically there isn't a good way to get the Xbox Button functionality on the Ally.  I'm still messing around with REWASD to see if I can get another controller button (besides the Asus ones) to trigger the home button where it works good.

**Reply by seerge** [reply]
I don't see an easy way to imitate xbox controller from the app and even if it will be able to do so it will be "virtual" one (so for game it would look like a second controller connected to the system)

If the only thing you want from Xbox button is to open steam - why not to bind some keyboard button to that action in steam itself ?

**Reply by TheEnab1er** [reply]
Okay thanks for checking.  It's mainly when people are in handheld and on the go, they wouldn't have a keyboard to be able to bring it up.  It looks like REWASD does let you program a "delay" on press of the start button allowing it to trigger the Xbox button so that should be a great work around.  Thank you again for taking time to look into it.

**Reply by pinecones-sx** [reply]
Serge, one question I have, are the actual key values for the ROG controller buttons presenting as an Xbox controller or are they generic gamepad buttons where  the Crate software is responsible for remapping them to present as Xbox controller buttons?

I think this is part of the reason for these asks -- we're assuming that out of the box, the buttons present as actual Xbox controller buttons.

if that's the case, imo your software should support remapping hard button actions to Xbox buttons.  but I completely understand not wanting to be responsible for the feature creep and having to support the feature when third party resources exist to solve the problem.

**Reply by TheEnab1er** [reply]
Hey, from my testing they don't register as any type of controller at all, nor any type of input device.  REWASD doesn't detect ANY type of input from them.  In USB Game Controllers there is 0 input, nothing in device manager, etc.  They're custom keys put in by Asus that (from my understanding) run at a BIOS level.  So you'd have to create a program that could take hold of those values and turn them into functions.  

Seerge will know more, but basically there isn't a simple re-map with the Asus buttons into controller inputs.   G-Helper does allow mapping to the virtual windows key input mappings, but I'm assuming a controller is in an entirely different realm of like he said needing to create a virtual 2nd controller.

**Reply by seerge** [reply]
I don't have Ally anymore. I'v returned it back to store, cause didn't really use it. But as I remember Ally controller identifies itself (either in device manager or in driver-downloads) as ROG Raikiri.

Most probably it emulates all XBox stuff on hardware level (after all you can use Ally in most games w/o any software at all, just with drivers)

The only difference is "asus" special buttons I mentioned in previous comments, those are not part of controller at all, and come in separately and need to be caught by g-helper (or AC services)

Remapping whole controller (and for what purpose?), means intentionally listening and passing through all controller input, and then emulating something else from that. That's a huge task, that is solved by specialized or commercial apps like REWASD or others. And I (as a single developer who does all this in a free time) don't have resources to do everything :)

**Reply by seerge** [reply]
Quick question (since I don't have Ally anymore). Can you map M1 and M2 in AC separately to different actions ? or they always act as same button?

If you can - can you try this build (and press M1 and M2 and check what Key they put in app log ? )

Thanks
[GHelper.zip](https://github.com/seerge/g-helper/files/12302483/GHelper.zip)

**Reply by TheEnab1er** [reply]
Yes, you can map the buttons to other keys in Armoury Crate.  However I found if the "Armoury Crate SE Service" isn't running then the back paddle maps I had in AC don't work.  However using G-Helpers built in "Stop" services leaves that service running (which is a good thing) because then the back paddles still hold the maps people set in AC.  And the ROG / Command Center button don't open Armoury Crate / Command Center since "Asus Optimization" is disabled (which is good and I don't think I'd change anything there).

I'm still showing WMI event 165 for both in the log if they haven't been set in Armoury Crate (default settings).  Then I mapped the back paddles (left one LB / right one RB) and they don't show up at all in G-Helper when pressing the keys.  Even setting something for "Back Paddles" like "Mode" no longer functions since they're setup as bumpers inside Armoury Crate.

Hope the info helps :)

**Reply by seerge** [reply]
Oke, then most probably this "mapping" happens on hardware level. I.e. they just start emitting whatever key you have mapped them in AC . Do I understand correctly that custom mappings work even when AC SE Service is stopped ?

**Reply by TheEnab1er** [reply]
The mapping does not work if AC SE is disabled.  In G-Helper they still work as a single button, but if I set my back buttons to something inside of AC (Like "A", or "B" or "RB"), the "AC SE Service" has to be ON for them to work.  

So right now if I click "Stop" in G-Helper it works great, because it stops the ROG button from opening Armoury Crate (so I can use it to open G-Helper) and it stops the Command Center button from opening Command Center (so I can map it in G-Helper to power profile switcher), and my mappings on the back paddles still work to whatever buttons I assigned them inside AC.

**Reply by TheEnab1er** [reply]
Also thank you for all the updates.  A lot of my viewers love the program and are using it over Armoury Crate :) (including myself).  It's awesome, great work!

**Reply by seerge** [reply]
Ok, it's still strange that AC somehow manages to distinguish M1 and M2, despite them reporting same Key ID

**Reply by seerge** [reply]
@TheEnab1er hey, quick question to you as a power user. I see that you still use G-Helper (even in combo with AC). 

I have recently found a way to add a toggle between gamepad / mouse-emulation mode for Ally. And was thinking on other features.

Main problem is that AC on Ally does A LOT MORE specific stuff, than AC does on desktops. So there is  significant gap in terms of functionality between AC on Ally and G-Helper.

And for me (as a single person) manually implementing everything that AC does there, is almost impossible. But I was wondering what are the main thing that hold you to AC ? Game Profiles ? Gyro ? Controller Calibration ? Custom M1/M2 key bindings?

P.S. Thanks for featuring app again in the video ;)

**Reply by TheEnab1er** [reply]
@seerge Hey.  The only reason I use AC is to be able to use the back paddles in games.  I personally use reWASD for all the desktop/mouse mappings as I have a bunch of shortcuts setup to alt-tab, force close, etc.  However even reWASD has no mapping for the back paddles on the Ally.  

The only other thing I like about having AC enabled is access to command center.  It makes changing brightness, toggling CPU Boost, refresh rate, etc faster.  I think for most people it's command center.  If G-Helper had a way to pop up with a more touch friendly overlay without minimizing the game it'd be a lot more useful on the Ally IMO.  

But you've already done more than enough, it's an awesome app and works great.  A lot of people are still just using AC so I wouldn't want you to put a bunch of time into it.   I've been using the Steam Deck OLED a lot more than the Ally lately anyway, just for the superb battery life.  

I've been thinking about getting another G14 (RTX 4090) and I'll definitely be using G-Helper on a clean wipe if I do, keep up the great work :)

**Reply by TheEnab1er** [reply]
Hey, I saw the new update with the back button support and MUCH more, awesome work!!  I already made a video about it, and I included a couple things that could added and be of great value since the back buttons/re-mapping is working: https://youtu.be/2aEA-Varn-8

**Reply by seerge** [reply]
@TheEnab1er hello, thanks for a quick video and feedback

I have watched it and here are my comments

1. **Custom FPS limit** - possible, but I would prefer to keep UI simple. Based from what I have seen in AMD settings ``120 fps limit`` = ``no limit``. I can add 45FPS and 90FPS in next update

2. **Stopping AC service** - I can add it to the list, but whole idea about the app, is that it works as a replacement for Armoury (you use it on a laptop so you know). So it's not really designed to "fight" with asus stuff anyhow. Currently stoppable services, are the ones that will be running even on a clean windows install (they just come auto-installed along Asus Control Interface driver). 

3. **Deadzones / Vibra / Bindings** - they are all also native hardware settings and not emulated anyhow (unlike other apps do that). So G-Helper just sets same thing as AC can via same protocol (so 0-100% deadzone in G-Helper should be same as in AC.) You can even quick app, and your mouse mode or bindings will keep working until something else overwrites them or you put device to sleep / reboot.

4. **Controller Calibration** - is something I can just add in future as well :) It will depend if current updates will be picked up by community or not

5. **Bindings for Command Center / ROG buttons** - unfortunately, unlike any other buttons, those work differently. They work same as M1,M2,M3,M4 buttons on a laptops. So you can't bind M+ROG combo (secondary action) to them at all. And you also can't bind them to actual controller actions (like XBox key). I just set Control center to toggle performance modes, and ROG - to open G-Helper, and call it a day.

6. **Custom hotkey combinations (or macros)** for bindings - that's something I want to add in future. So ``Alt+F4`` or ``Alt+Tab`` should be possible. Hopefully I will  find a way to do brightness controls via hotkeys as well :)

**Reply by TheEnab1er** [reply]
1.  Ahh that makes sense and keeps it easy/intuitive to use. 

2. What I was meaning here is in the Extra tab if you click "Stop" to stop the Asus services, it leaves "ARMOURY CRATE SE Service" running.  Which directly conflicts with the Auto, Mouse, Gamepad mode in G-Helper.  So the Command Center button and ROG button don't open any Asus software, but there is that SE Service running in the background causing "Auto" mode to stop working as soon as any other window is clicked besides G-Helper.

I think the reason for this is when the services are enabled, you'll see a pop up at the top left over and over where it says "Gamepad" mode.  It seems like Armoury Crate is refreshing the controls anytime an active window is changed.  And while the message isn't popping up, that "ARMOURY CRATE SE Service" is refreshing and overriding G-Helper.  Which is why I showed how to disable in the video, then G-Helpers controls work perfect and switch back and forth. 

3. These seem to work good :) 

4. That'd be great to just have the trigger calibration added as every Ally I've had (3 of them) has had triggers that don't go to 100%.  It's something with the Hall Effect and the sensor.

5. Just to make sure, it's not possible to make it where the ROG button activates the secondary functions?  So if I assigned "D-Pad Up" to "Volume Up" in the Secondary section, is there a way that instead of having the back paddles be M1 or M2, I could have the ROG button be M1 or M2?  Then I could hold the ROG button and press up on the D-Pad to adjust the volume instead of loosing one of the back paddles.  This would be handy so both paddles can be used in games, but then there is the ROG button to activate the Secondary layer.

6. Sweet!  Those would be great shortcuts to have :) 

And thanks for all the support on the program.  I've already disabled AC completely as I have no need for it now.  And I do use G-Helper on my Asus laptops as well as it works perfect 👌

**Reply by seerge** [reply]
@TheEnab1er 

> What I was meaning here is in the Extra tab if you click "Stop" to stop the Asus services, it leaves "ARMOURY CRATE SE Service" running.

I can add AC services to the Stop / Start list (it's easy). But don't you think people will start screaming online that G-Helper destroys armoury installation ? (i know it's nonsence, but I can imagine this happening :))

> Just to make sure, it's not possible to make it where the ROG button activates the secondary functions? So if I assigned "D-Pad Up" to "Volume Up" in the Secondary section, is there a way that instead of having the back paddles be M1 or M2, I could have the ROG button be M1 or M2? 

Unfortunately that's not possible, both Primary and Secondary actions are handled by Ally's controller on hardware level. 
So you can "set" what command to emit while actual button is pressed (i.e. DPad Up) and what to emit when M1/M2+Dpad Up is pressed. 

So pretty much same thing as AC can bind - is also possible to implement, but not more :)

**Reply by seerge** [reply]
@TheEnab1er, hello. I'm now preparing a new update, and I have added some changes (also according to your feedback)

- New real-time battery discharge reading (from same source armoury's overlay gets it)
- FPS limiter settings ``30`` - ``45`` - ``60`` - ``90`` - ``120`` - ``OFF``
- Controller bindings now work in all modes (including Mouse / Desktop)
- ``M + DPad Lefft/ Right`` to control brightness
- Added new binding options such as ``Close Window``, ``Alt-Tab``, ``Brightness Up / Down``, ``Toggle AMD Overlay``, etc
- Stop / Start Services button under ``Extra`` now also suspends Armoury Crate main service (if you don't want it to interfere with controller modes)

[GHelper.zip](https://github.com/seerge/g-helper/files/13997751/GHelper.zip)

And since it's now possible to do bindings in any mode, I wanted to "tweak" default bindings a bit . For example I don't see much sense in all secondary bindings for M+DPad / M+ABXY that were showing some notification center (who is even using that?), or launching video recording, etc.

So question to you (as a power user), what would you prefer M+Dpad / M+ABXY bindings do ? Or even turn them off by default ?

Current bindings are simple

DpadUP/DOWN - brightness
M+Y - AMD overlay toggle

**Reply by bullghost** [reply]
@seerge. hi i  was using g-helper for replacing the armoury crate, beside tdp , i use it as main app for controller mapping,
i found when on the desktop mode /mouse mode M1 + DPadUP didnt bringup softkeyboard,
and the most important or missing for me is disable / enable embedded controller function.
oh one more on dropdown list on key binding in extra setting its will be good to have xbox button to choose so we can map it to one of soft menu key.

i have test your GHelper that you attached here, i have notice a block square area on left top corner sometimes flickering at the time i test i have amd adrenaline fps overlay box on the right top corner. but with GHelper release version its didnt happen.

So question to you (as a power user), what would you prefer M+Dpad / M+ABXY bindings do ? Or even turn them off by default ?
i try answer from my perpective view ,  turn of by default, , 
"i prefer M+Dpad down for task manger should be default" <- this one by default its mapped , its maybe related to its driver where the driver was XboxController from asus.

my setting :
clean windows install, no armoury crate, with native amd ryzen driver and adrenaline ,

thankyou so much for awesome work , finally we can use g-helper on rog ally.

**Reply by seerge** [reply]
@bullghost hello

> disable / enable embedded controller

What is the practical scenario for that ?

> M1 + DPadUP didnt bringup softkeyboard,
[Edit] I have managed to find a way to show "Touch Keyboard" in windows programmatically

> oh one more on dropdown list on key binding in extra setting its will be good to have xbox button to choose so we can map it to one of soft menu key.

You can't bind XBOX button to anything outside of actual "controller" buttons (i.e. ROG/Control center buttons can't do that). Apparently it's firmware / hardware limitation as Armoury can't do that neither

> i have test your GHelper that you attached here, i have notice a block square area on left top corner sometimes flickering at the time i test i have amd adrenaline fps overlay box on the right top corner. but with GHelper release version its didnt happen.

I didn't get what you mean? what is "block square area on left top corner" ?

**Reply by seerge** [reply]
Build where 

``M`` + ``DPad+Up`` shows keyboard
``M`` + ``DPad+Down`` shows desktop
``M`` + ``DPad+Left / Right`` Controls Brightness
``M`` + ``Y`` - AMD overlay
``M`` + ``X`` - Screenshot

You can still override change this bindings under ``Controller ``

[GHelper.zip](https://github.com/seerge/g-helper/files/14001318/GHelper.zip)

**Reply by bullghost** [reply]
@seerge 

> What is the practical scenario for that ?

its use for when in docking mode , when using other joystick  with disabling the embedded controller will make other joystick to be recognized as player 1.

> I didn't get what you mean? what is "block square area on left top corner" ?

i mean there is blinking around top left corner the size is like the square of amd overlay
but your current GHelper doesnt having blinking square , i donno, i wait for it to come out but it didnt, 

your GHelper tested so far ,
it was true the M + DPad down was registered as open up task manager , it was default mapping from joystick driver,
i found this, after finish install all driver vga, controller and gyro driver, without any other app installed, i was able to open the task manager by pressing M + DPad Down

for me, register M + X  will accidently pressed, because when the game is so intense the M button got pressed some time, at armoury crate i have to disable all the  M +XYAB button, becuase its so ennoying, got pressed screen recording, screenshot, and another one its sent me to desktop.
for me the DPad is safer because most game i use analog stick for navigation

YAY  now we have button bringup the softkeyboard 👍

**Reply by seerge** [reply]
@bullghost  

> its use for when in docking mode , when using other joystick with disabling the embedded controller will make other joystick to be recognized as player 1.

I see, currently I don't know how to do that (besides manually disabling device in device manager, which I don't consider a good idea)

> it was true the M + DPad down was registered as open up task manager , it was default mapping from joystick driver,

But is handy mapping ? How often do you use that ? 

Personally i rarely kill tasks via task manager :) since I have to "recreate" all mappings from scratch (due to the way mapping work on Ally), we are free to change things. And my question what would be optimal default settings :)

**Reply by bullghost** [reply]
@seerge 

> I see, currently I don't know how to do that (besides manually disabling device in device manager, which I don't consider a good idea)

yes i agree with you.

>  And my question what would be optimal default settings

iam not sure for this, maybe others not using this function, for me its might be handy if anything going wrong in the game like the game is freezed  or something, you can come to task manager to kill it ,

**Reply by bullghost** [reply]
@seerge i found this for power profile. it might be useful
Turn off CPU Boost technology for better performance in some games
The switch for this feature is currently implemented on version 1.4.5 of the RC Control Center and is called the CPU Boost switch.
This feature is also available on the somatosensory assistant, called Turbo Switch.
25W or 30W is the maximum power limit of the APU, and some games that tend to have high GPU utilization will suffer if CPU Boost (CPU acceleration technology) is enabled by default, which is also known as GPU bound application, because most of the power of the whole machine is occupied by the CPU, in order to reduce the CPU power, free up excess power space to accommodate the GPU power, that is, to increase the proportion of GPU usage, the strategy is to turn off CPU Boost.

Refer to the login table below and set Attributes to 2 to enable the Processor Performance Boost Mode setting in Power Options:
`Windows Registry Editor Version 5.00

; 開啟 CPU Boost 選項顯示 (處理器效能提升模式)
[HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Power\PowerSettings\54533251-82be-4824-96c1-47b60b740d00\be337238-0d82-4146-a960-4f3749d470c7]
"Attributes"=dword:00000002`

Next, select a power plan under "Power Options" (Search On), with Turbo being 25/30W Extreme, Performance being 15W and Silent being 10W. Select the plan you want to change, then click "Change Plan Configuration" - > "Change Advanced Power Configuration" - > expand "Processor Power Management" - > expand "Processor Performance Boost Mode" - > change "Battery Active" and "Normal Power" to "Deactivated", and recommend that Turbo and Performance and Silent are deactivated. powercfg.cpl

found it on hackmd.io

**Reply by seerge** [reply]
@bullghost CPU Boost is available in the app already, for ages. It's under ``Fans + Power``.

**Reply by TheEnab1er** [reply]
@seerge Hey, I'd mainly use shortcuts that are already keyboard keys that can be bound.  I do a lot of testing to having benchmark shortcuts as secondary is what I use most and I assign those to keyboard keys.  

Adding in the alt+f4, brightness, etc is what I think was missing for most.  The only other thing that could be worth adding is a button to turn off G-Helpers controller mode (Auto/Gamepad/Desktop) since some people will want to use AC for the custom game profiles and command center.  

Also I saw someone mentioned about disabling the embedded controller.  It's a key functionality of Command Center as it allows you turn off the internal controls.  Without that option it's basically like having a controller plugged in via usb permanently without being able to turn it off.  And since people dock the Ally it's something that would help a bit.

And I personally use reWASD most of the time because of the sheer control I have over everything.  I use a lot of combo keys, shift layers, etc, to control everything on the machine.  It's also why the embedded controller being on isn't a big deal to me since I've combined the embedded and external controller into 1.  So I can turn a controller on while playing and it still appears as just 1 controller everywhere.

Thanks for all the updates :). It's already good enough that I don't use AC at all anymore.

**Reply by seerge** [reply]
@TheEnab1er 

> The only other thing that could be worth adding is a button to turn off G-Helpers controller mode (Auto/Gamepad/Desktop) since some people will want to use AC 

It's already possible in a build linked here (it's 4th mode called "Skip", it will just completely skip any controller-related things)

**Reply by seerge** [reply]
Version posted here is live as https://github.com/seerge/g-helper/releases/tag/v0.149

**Reply by seerge** [reply]
@bullghost @TheEnab1er 

I have managed to find a way to "disable" XBox controller same as AC does it (there is a special command that can be sent to it), so it's now live along with other updates in latest
https://github.com/seerge/g-helper/releases/tag/v0.150

**Reply by seerge** [reply]
P.S. I have also found a way to bind **left** stick to mouse-movement instead of **right**. Is it something that you guys think can be useful ?

**Reply by TheEnab1er** [reply]
Awesome!  In the quick test and the disable controller function seems to be working well 👍.   I think most people will want the right stick to be the mouse since that's how AC does it.  However, is there a way to set the Secondary Function as a Mouse for the LStick?  So when pressing M1 or M2, it'll temporarily turn it into a mouse.  

This can be really useful for quickly accessing a mouse in games, or if in gamepad mode there isn't a need to switch to "Mouse" mode to be able to navigate around.

**Reply by seerge** [reply]
@TheEnab1er 

It's not possible to make it as a secondary function - as M1 / M2 produce keycode only when pressed, not when unpressed, so I simply don't know when to turn it off.

But it's possible to add a "binding" that would turn it ON or OFF. And this is already added in a test build in this thread
https://github.com/seerge/g-helper/discussions/1985#discussioncomment-8274395

**Reply by TheEnab1er** [reply]
Hey, I just bought a G14 (RTX 4060) and your app is 👌 on it.  Even the function lock, being able to program the keys, etc.  The "Ultimate" icon changing red is a nice touch as well to know it's in dGPU mode.  Anyway, you've made one of the best apps for Asus laptops/products.  I'm sure it makes a huge difference to the people that use it as AC is not even close to what you've made.  Appreciate it man!

**Reply by seerge** [reply]
@TheEnab1er amazing :) you can also overclock your nvidia GPU a bit, cause this is almost "free" performance gain

**Reply by TheEnab1er** [reply]
Hey Seerge, G-Helper has been working great on the G14 😊.  I was able to fully overclock the GPU to +250Mhz Core and +500 Mem stable.  Then I tried MSI just to see and it started crashing at around 280, so 250 is perfect and I like the G-Helper OC way more than MSI.

I was wondering if it's possible to add something besides 60Hz refresh rate when on battery?  165Hz uses a lot of battery, but when working on the computer 60Hz is unbearably laggy, especially when using the laptop touchpad.  Something like 90Hz would be a great in-between.  Or maybe even 82.5Hz since it's exactly half the 165Hz.  I saw Windows has a Dynamic mode that will do 82.5Hz or 165Hz (depending on motion), but it messes with some games so it's not worth using. 

Honestly this isn't important at all though and I'm sure most people don't really care about it.  I just wanted to ask in case it was really easy to modify.  If it requires any time or you don't think it'd be something worth adding then don't worry about it at all as everything else is working great :)   Appreciate the work you do man!

**Reply by seerge** [reply]
@TheEnab1er hello :)

G-Helper can set only refresh rates already available in windows (as buttons in the app is just a shortcuts to same thing in default settings screen)

So if your device has only 60 and 165, you can't set anything else.

You can try to tinker with https://www.monitortests.com/forum/Thread-Custom-Resolution-Utility-CRU and try to create a custom refresh rate, and set it manually. But I would not bother.

In my experience impact from 60Hz is actually minimal (mb something like 0.5W-1W of discharge). So I'm usually sacrificing it in favor of smoother UI.

**Reply by seerge** [reply]
@TheEnab1er btw, since you messaged in Ally thread.

I have prepared an Experimental version of Auto TDP for Ally. It's very basic, just adjusts power limits to match currently set FPS limit in AMD driver.

TDP is adjusted in range from 6W up to what your currently active mode defines

But you can try it if you want (I will push it live soon as well)

### [GHelper.zip](https://github.com/seerge/g-helper/files/15134579/GHelper.zip)

**Reply by TheEnab1er** [reply]
Ahh gotcha, that makes sense.  And 100%, I just keep mine on 165Hz most of the time.  The only time it matters more is when watching a movie/show as 60Hz saves a decent amount over the course of a few hours.

**Reply by TheEnab1er** [reply]
Sweet!  That sounds similar to Auto TDP.  I haven't messed with the Ally much at all lately, but this sounds pretty interesting and I'll give it a shot 👍

**Reply by TheEnab1er** [reply]
Just tried it out, this works way better than another "Auto TDP" program I tried a while back.  It doesn't slow down near as much when going from a low demand area to a high demand area quickly. 

I'll be testing a variety of games later including lower demanding games to see how far it pulls it back with very low demand.

I did notice it pulling the FPS back a bit further than the FPS limit, even if just standing still.  So if it's set to 60FPS, it's frequently in the low-mid 50's (at least in Crysis Remastered).  However, I think that's preferable for the extra power saving.  With the VRR display it shouldn't affect gameplay much.  Also it's really stable in the frametime graph and doesn't have hardly any wavy lines/slow downs which is the most important thing.

Overall it works great from my initial testing and you've done a superb job, thanks a bunch from the whole Ally community.👍  I'm wanting to make a video and plan to, but I just haven't had much time to work on YouTube stuff.

**Reply by seerge** [reply]
> I did notice it pulling the FPS back a bit further than the FPS limit, even if just standing still. So if it's set to 60FPS, it's frequently in the low-mid 50's (at least in Crysis Remastered).

It is. Current logic is simple

FPS > 90% of limit  constantly for 3 seconds without drops => slow down TDP if there is a headroom for that
FPS < 80% of a limit even once => immediately increase TDP

So it aims at 90% of your target FPS.

It's tricky to set a bar higher, cause then margin between actual limit (let's say its 60FPS) is just too small to actually detect if it's stable enough

**Reply by TheEnab1er** [reply]
That makes sense and I think it's better to have the FPS stay slightly lower and for it to be stable than to try and hold 60 (or target FPS).  It's very well implemented to where it's usable even going from high demand to low demand.   

I had never thought of actually using AUTO TDP because the other program I used had very bad dips/slow downs.  And it's probably from exactly what you said, no margin available.  But the way you've implemented it is great.  

I'm excited to test it out more later when I have time :)

**Reply by seerge** [reply]
Sure. Anyway all this logic is obviously not final and I can tweak / adjust it later.
I just wanted to start with something very simple and see if it works.

I tried GTA5, Death's Door, Dead Cells (runs on 6w!), Tomb Raider and few other games. Everywhere it seem to work.

It obviously won't work well, if AMD's default FPS limiter is not working for that game (for example it runs in a window or so) or if it's not possible to read FPS from that game.

**Comment by TheEldritchVoid** [comment]
Hey so another ally user here and it's almost completely broken over here, the biggest quality of life bug I'm running into with any solution is the popup to download the native software when I hit those two buttons

**Reply by seerge** [reply]
@TheEldritchVoid hello, your question [is answered in the very first FAQ answer.](https://github.com/seerge/g-helper/wiki/FAQ#how-do-i-stop-the-armoury-crate-install-popup-appearing-every-time-i-press-the-m4--rog-key)
